### PR TITLE
Implement tensor.select(Dimname,int)

### PR DESCRIPTION
**Pull request by zou3519**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #21919 Added names= arg for some creation functions
* #22077 Refactor function_wrapper.create_generic
* **#21795 Implement tensor.select(Dimname,int)**

tensor.select(Dimname, int) wraps tensor.select(int,int), where the
named inference rules happen.

Codegen changes:
- named tensor code in core/{Type.h, Tensor.h, TensorMethods.h} is
guarded with macros.

Test Plan
- `python test/test_namedtensor.py -v` [namedtensor ci]

gh-metadata: pytorch pytorch 21795 gh/zou3519/54/head

Differential Revision: [D15833457](https://our.internmc.facebook.com/intern/diff/D15833457)

### CODE REVIEW

**Review by nairbv — APPROVED** [review]

File: tools/jit/gen_jit_dispatch.py [hunk 1]
```diff
@@ -171,7 +171,7 @@ def from_ivalue(arg, value):
 """)
 
 
-blacklisted_types = {'Storage', 'Dimname', 'DimnameList?'}
+blacklisted_types = {'Storage', 'DimnameList?', 'ConstQuantizerPtr', 'Dimname', 'DimnameList?'}
```
Comment (nairbv) [inline tools/jit/gen_jit_dispatch.py]: DimnameList twice here?

Comment (zou3519) [reply tools/jit/gen_jit_dispatch.py]: Nice catch

**Comment by zou3519** [comment]
@ailzhang XLA failure. Please let me know how you would like to proceed.

**Comment by ailzhang** [comment]
Hi @zou3519 Thanks! I think this will be a simple schema change at the first glance. We should be able to make a followup change once this is merged. cc: @dlibenzi @asuhan

**Comment by dlibenzi** [comment]
> Hi @zou3519 Thanks! I think this will be a simple schema change at the first glance. We should be able to make a followup change once this is merged. cc: @dlibenzi @asuhan

Looks like the code which is generating `RegistrationDeclarations.h` needs fixing?

**Comment by zou3519** [comment]
> Hi @zou3519 Thanks! I think this will be a simple schema change at the first glance. We should be able to make a followup change once this is merged. cc: @dlibenzi @asuhan

I'm not sure if it's a schema change; I think you just need to blacklist Dimname/DimnameList in the code generation

**Comment by dlibenzi** [comment]
> > Hi @zou3519 Thanks! I think this will be a simple schema change at the first glance. We should be able to make a followup change once this is merged. cc: @dlibenzi @asuhan
> 
> I think you just need to blacklist Dimname/DimnameList in the code generation

Why blacklisting?
Can you post the `RegistrationDeclarations.h` generated by your PR?
Or at least the incriminated line?

**Comment by zou3519** [comment]
> Why blacklisting?
Can you post the RegistrationDeclarations.h generated by your PR?
Or at least the incriminated line?

https://gist.github.com/zou3519/40ef0f94399d53c8996720bec852908f
I think the problem is that the NAMEDTENSOR_ENABLED macros aren't being respected somewhere downstream. I could change it so that everything that is currently wrapped inside the NAMEDTENSOR_ENABLED sections are not written to the RegistrationDeclarations.h file if that is the only file being consumed.

**Comment by dlibenzi** [comment]
> > Why blacklisting?
> > Can you post the RegistrationDeclarations.h generated by your PR?
> > Or at least the incriminated line?
> 
> https://gist.github.com/zou3519/40ef0f94399d53c8996720bec852908f
> I think the problem is that the NAMEDTENSOR_ENABLED macros aren't being respected somewhere downstream. I could change it so that everything that is currently wrapped inside the NAMEDTENSOR_ENABLED sections are not written to the RegistrationDeclarations.h file if that is the only file being consumed.

I think, if XLA should not generate code for it, the #ifdef NAMEDTENSOR_ENABLED shouldn't even be included in that file

**Comment by zou3519** [comment]
> > > Why blacklisting?
> > > Can you post the RegistrationDeclarations.h generated by your PR?
> > > Or at least the incriminated line?
> > 
> > 
> > https://gist.github.com/zou3519/40ef0f94399d53c8996720bec852908f
> > I think the problem is that the NAMEDTENSOR_ENABLED macros aren't being respected somewhere downstream. I could change it so that everything that is currently wrapped inside the NAMEDTENSOR_ENABLED sections are not written to the RegistrationDeclarations.h file if that is the only file being consumed.
> 
> I think, if XLA should not generate code for it, the #ifdef NAMEDTENSOR_ENABLED shouldn't even be included in that file

Are there other files I should be aware of that XLA uses to generate code, or is this (RegistrationDeclarations) the only one?

**Comment by dlibenzi** [comment]
> > > > Why blacklisting?
> > > > Can you post the RegistrationDeclarations.h generated by your PR?
> > > > Or at least the incriminated line?
> > > 
> > > 
> > > https://gist.github.com/zou3519/40ef0f94399d53c8996720bec852908f
> > > I think the problem is that the NAMEDTENSOR_ENABLED macros aren't being respected somewhere downstream. I could change it so that everything that is currently wrapped inside the NAMEDTENSOR_ENABLED sections are not written to the RegistrationDeclarations.h file if that is the only file being consumed.
> > 
> > 
> > I think, if XLA should not generate code for it, the #ifdef NAMEDTENSOR_ENABLED shouldn't even be included in that file
> 
> Are there other files I should be aware of that XLA uses to generate code, or is this (RegistrationDeclarations) the only one?

We also use `Functions.h` and `NativeFunctions.h`, but you can leave those alone.
The real driver for the code generation is `RegistrationDeclarations.h`.

https://github.com/pytorch/xla/blob/master/scripts/generate_code.sh

**Comment by dlibenzi** [comment]
> > > > > Why blacklisting?
> > > > > Can you post the RegistrationDeclarations.h generated by your PR?
> > > > > Or at least the incriminated line?
> > > > 
> > > > 
> > > > https://gist.github.com/zou3519/40ef0f94399d53c8996720bec852908f
> > > > I think the problem is that the NAMEDTENSOR_ENABLED macros aren't being respected somewhere downstream. I could change it so that everything that is currently wrapped inside the NAMEDTENSOR_ENABLED sections are not written to the RegistrationDeclarations.h file if that is the only file being consumed.
> > > 
> > > 
> > > I think, if XLA should not generate code for it, the #ifdef NAMEDTENSOR_ENABLED shouldn't even be included in that file
> > 
> > 
> > Are there other files I should be aware of that XLA uses to generate code, or is this (RegistrationDeclarations) the only one?
> 
> We also use `Functions.h` and `NativeFunctions.h`, but you can leave those alone.
> The real driver for the code generation is `RegistrationDeclarations.h`.
> 
> https://github.com/pytorch/xla/blob/master/scripts/generate_code.sh

We could pass that through `cpp` on our side if you like.

**Comment by zou3519** [comment]
> > > > > > Why blacklisting?
> > > > > > Can you post the RegistrationDeclarations.h generated by your PR?
> > > > > > Or at least the incriminated line?
> > > > > 
> > > > > 
> > > > > https://gist.github.com/zou3519/40ef0f94399d53c8996720bec852908f
> > > > > I think the problem is that the NAMEDTENSOR_ENABLED macros aren't being respected somewhere downstream. I could change it so that everything that is currently wrapped inside the NAMEDTENSOR_ENABLED sections are not written to the RegistrationDeclarations.h file if that is the only file being consumed.
> > > > 
> > > > 
> > > > I think, if XLA should not generate code for it, the #ifdef NAMEDTENSOR_ENABLED shouldn't even be included in that file
> > > 
> > > 
> > > Are there other files I should be aware of that XLA uses to generate code, or is this (RegistrationDeclarations) the only one?
> > 
> > 
> > We also use `Functions.h` and `NativeFunctions.h`, but you can leave those alone.
> > The real driver for the code generation is `RegistrationDeclarations.h`.
> > https://github.com/pytorch/xla/blob/master/scripts/generate_code.sh
> 
> We could pass that through `cpp` on our side if you like.

What are you referring to by "that"?

I fixed up RegistrationDeclarations.h in this pull request so that it doesn't emit any `#ifdef NAMEDTENSOR_ENABLED` code and it looks like the build succeeded so all should be good.

**Comment by dlibenzi** [comment]
I meant we could pipe that through the C++ preprocessor before our parsing
... but if you already fixed that, we are good

**Pull request closed without merging**